File: 2024 January/Distribute candies in a binary tree/main.py
# Link: https://www.geeksforgeeks.org/problems/distribute-candies-in-a-binary-tree/1


class Node:
    def __init__(self, val):
        self.right = None
        self.data = val
        self.left = None


class Solution:
    def distributeCandy(self, root):
        self.moves = 0

        def dfs(root):
            if not root:
                return 0

            left_requires = dfs(root.left)
            right_requires = dfs(root.right)

            self.moves += abs(left_requires) + abs(right_requires)
            remaining_coins = left_requires + right_requires + root.data - 1

            return remaining_coins

        dfs(root)
        return self.moves


if __name__ == "__main__":
    obj = Solution()
    root = Node(3)
    root.left = Node(0)
    root.right = Node(0)
    print(obj.distributeCandy(root))
    root1 = Node(0)
    root1.left = Node(3)
    root1.right = Node(0)
    print(obj.distributeCandy(root1))

File: 2024 January/Distribute candies in a binary tree/main.cpp
// Link: https://www.geeksforgeeks.org/problems/distribute-candies-in-a-binary-tree/1

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int key;
    Node *left, *right;
};

class Solution
{
public:
    int ans = 0;

    int help(Node *root)
    {
        if (root == NULL)
        {
            return 0;
        }
        if (root->left == NULL && root->right == NULL)
        {
            return root->key - 1;
        }
        int lf = help(root->left);
        int rg = help(root->right);
        ans += abs(lf);
        ans += abs(rg);
        return lf + rg + root->key - 1;
    }

    int distributeCandy(Node *root)
    {
        int k = help(root);
        return ans;
    }
};

int main()
{
    Solution obj;
    Node *root = new Node();
    root->key = 3;
    root->left = new Node();
    root->left->key = 0;
    root->right = new Node();
    root->right->key = 0;
    cout << obj.distributeCandy(root) << endl;
    Node *root1 = new Node();
    root1->key = 0;
    root1->left = new Node();
    root1->left->key = 0;
    root1->right = new Node();
    root1->right->key = 3;
    cout << obj.distributeCandy(root1) << endl;
    return 0;
}